sa/sa.go

```diff
@@ -915,3 +915,188 @@ func (ssa *SQLStorageAuthority) Health(ctx context.Context) error {
+// If the request names a specific shard, this function will return an error
+// if that shard is already leased. Otherwise, this function will return the
+// index of the oldest shard for the given issuer.
+func (ssa *SQLStorageAuthority) LeaseCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	if core.IsAnyNilOrZero(req.Until, req.IssuerNameID) {
+		return nil, errIncompleteRequest
+	}
+	if req.Until.AsTime().Before(ssa.clk.Now()) {
+		return nil, fmt.Errorf("lease timestamp must be in the future, got %q", req.Until.AsTime())
+	}
+
+	if req.MinShardIdx == req.MaxShardIdx {
+		return ssa.leaseSpecificCRLShard(ctx, req)
+	} else {
+		return ssa.leaseOldestCRLShard(ctx, req)
+	}
+}
+
+// leaseOldestCRLShard finds and then leases the crl shard for the given issuer
+// which is not currently leased and hasn't been updated in the longest. It
+// returns an error if all shards for that issuer are already leased.
+func (ssa *SQLStorageAuthority) leaseOldestCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	shardIdx, err := db.WithTransaction(ctx, ssa.dbMap, func(txWithCtx db.Executor) (interface{}, error) {
+		var shards []*crlShardModel
+		_, err := txWithCtx.Select(
+			&shards,
+			`SELECT id, issuerID, idx, thisUpdate, nextUpdate, leasedUntil
+			FROM crlShards
+			WHERE issuerID = ?`,
+			req.IssuerNameID,
+		)
+		if err != nil {
+			return -1, err
+		}
+
+		// Convert the slice to a map, so we can easily detect shards that we expect
+		// to exist, but don't. At the same time, keep track of the oldest shard
+		// we've seen, so we can work on that one if none are missing.
+		shardMap := make(map[int]*crlShardModel, len(shards))
+		var oldest *crlShardModel
+		for _, shard := range shards {
+			if shard.Idx < int(req.MinShardIdx) || shard.Idx > int(req.MaxShardIdx) {
+				continue
+			}
+			shardMap[shard.Idx] = shard
+			if shard.LeasedUntil.After(ssa.clk.Now()) {
+				continue
+			}
+			if oldest == nil ||
+				(shard.ThisUpdate == nil && oldest.ThisUpdate != nil) ||
+				shard.ThisUpdate.Before(*oldest.ThisUpdate) {
+				oldest = shard
+			}
+		}
+
+		if oldest == nil {
+			return -1, fmt.Errorf("issuer %d has no unleased shards in range %d - %d", req.IssuerNameID, req.MinShardIdx, req.MaxShardIdx)
+		}
+
+		// Determine which shard index we want to lease: by default the oldest, but
+		// an arbitrary missing one if any are missing.
+		shardIdx := oldest.Idx
+		needToInsert := false
+		for i := req.MaxShardIdx; i >= req.MinShardIdx; i-- {
+			_, ok := shardMap[int(i)]
+			if !ok {
+				shardIdx = int(i)
+				needToInsert = true
+				break
+			}
+		}
+
+		if needToInsert {
+			_, err = txWithCtx.Exec(
+				`INSERT INTO crlShards (issuerID, idx, leasedUntil)
+				VALUES (?, ?, ?)`,
+				req.IssuerNameID,
+				shardIdx,
+				req.Until.AsTime(),
+			)
+		} else {
+			_, err = txWithCtx.Exec(
+				`UPDATE crlShards
```

Comment:
Add an explicit `LIMIT 1` since we only ever intend to have 1 row per issuerID and shard index pair.